core/domain/stats_jobs_one_off.py

```diff
@@ -16,16 +16,212 @@
+        """
+        # Handling state additions, deletions and renames.
+        for change_dict in change_list:
+            if change_dict['cmd'] == exp_domain.CMD_ADD_STATE:
+                old_stats_model.state_stats_mapping[change_dict[
+                    'state_name']] = (stats_domain.StateStats
+                                      .create_default())
+            elif change_dict['cmd'] == exp_domain.CMD_DELETE_STATE:
+                old_stats_model.state_stats_mapping.pop(change_dict[
+                    'state_name'])
+            elif change_dict['cmd'] == exp_domain.CMD_RENAME_STATE:
+                old_stats_model.state_stats_mapping[
+                    change_dict['new_state_name']] = (
+                        old_stats_model.state_stats_mapping.pop(
+                            change_dict['old_state_name']))
+        old_stats_model.exp_version += 1
+
+        return old_stats_model
+
+    @classmethod
+    def _recompute_statistics(
+            cls, exp_id, exp_version, events, old_stats_model):
+        """Create a ExplorationStatsModel with the statistics from
+        events added.
+
+        Args:
+            exp_id: str. The id of the exploration to recompute statistics for.
+            exp_version. The version of the exploration.
+            events: list(dict). A list of event data dicts from the datastore.
+            stats_model: ExplorationStatsModel. The stats_model to recompute
+                statistics for.
+        """
+        state_hit_ids, solution_hit_ids = set(), set()
+
+        for event in events:
+            state_stats = old_stats_model.state_stats_mapping[event['state']]
+            if event['type'] == 'answer_submitted':
+                state_stats.total_answers_count_v2 += 1
+                if event['is_feedback_useful'] is True:
+                    state_stats.useful_feedback_count_v2 += 1
+            elif event['type'] == 'state_hit':
+                state_stats.total_hit_count_v2 += 1
+                if event['session_id'] not in state_hit_ids:
+                    state_stats.first_hit_count_v2 += 1
+                    state_hit_ids.add(event['session_id'])
+            elif event['type'] == 'solution_hit':
+                if event['session_id'] not in solution_hit_ids:
+                    state_stats.num_times_solution_viewed_v2 += 1
+                    solution_hit_ids.add(event['session_id'])
+            elif event['type'] == 'exploration_actual_start':
+                old_stats_model.num_actual_starts_v2 += 1
+            elif event['type'] == 'complete_exploration':
+                old_stats_model.num_completions_v2 += 1
+            elif event['type'] == 'state_complete':
+                state_stats.num_completions_v2 += 1
+
+        for state, stats in old_stats_model.state_stats_mapping.iteritems():
+            old_stats_model.state_stats_mapping[state] = stats.to_dict()
+
+        stats_models.ExplorationStatsModel.create(
```

Comment:
These are too many put() operations. A better idea is to add all ExplorationStats dict objects to a list and use stats_models.ExplorationStatsModels.save_multi method.

Reply:
Done.